### PR TITLE
Wait for commit timeout before notifying consensus engine for availab…

### PR DESCRIPTION
**Pull request by b00f**
Fixing #3908

### CODE REVIEW

**Comment by b00f** [comment]
The log after this fix, Compare with #3908 log

```
notifyTxsAvailable:  size: 1, notifiedTxsAvailable: false
notifyTxsAvailable:  size: 2, notifiedTxsAvailable: true
notifyTxsAvailable:  size: 3, notifiedTxsAvailable: true
notifyTxsAvailable:  size: 4, notifiedTxsAvailable: true
notifyTxsAvailable:  size: 5, notifiedTxsAvailable: true
notifyTxsAvailable:  size: 6, notifiedTxsAvailable: true
notifyTxsAvailable:  size: 7, notifiedTxsAvailable: true
notifyTxsAvailable:  size: 8, notifiedTxsAvailable: true
notifyTxsAvailable:  size: 9, notifiedTxsAvailable: true
notifyTxsAvailable:  size: 10, notifiedTxsAvailable: true
I[2019-08-22|15:28:48.660] Executed block                               module=state height=6 validTxs=10 invalidTxs=0
I[2019-08-22|15:28:48.670] Committed state                              module=state height=6 txs=10 appHash=57169ECA1BDDF8D1D527D3FF3D7B5274B4D01902020783965DADBBE2FAE1AEF4
I[2019-08-22|15:28:53.673] Executed block                               module=state height=7 validTxs=0 invalidTxs=0
I[2019-08-22|15:28:53.677] Committed state                              module=state height=7 txs=0 appHash=57169ECA1BDDF8D1D527D3FF3D7B5274B4D01902020783965DADBBE2FAE1AEF4
```

File: mempool/clist_mempool.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ import (
 // mempool uses a concurrent list structure for storing transactions that can
 // be efficiently accessed by multiple concurrent readers.
 type CListMempool struct {
-	config *cfg.MempoolConfig
+	config *cfg.Config
```
Comment (melekes) [inline mempool/clist_mempool.go]: we must not require the full config here

File: mempool/clist_mempool.go [hunk 2]
```diff
@@ -446,17 +446,22 @@ func (mem *CListMempool) TxsAvailable() <-chan struct{} {
...
-		panic("notified txs available but mempool is empty!")
-	}
-	if mem.txsAvailable != nil && !mem.notifiedTxsAvailable {
-		// channel cap is 1, so this will send once
-		mem.notifiedTxsAvailable = true
-		select {
-		case mem.txsAvailable <- struct{}{}:
-		default:
+	go func() {
+		if mem.Size() == 0 {
```
Comment (melekes) [inline mempool/clist_mempool.go]: I think this check should be outside of the goroutine and it should really panic.

File: mempool/clist_mempool.go [hunk 3]
```diff
@@ -446,17 +446,22 @@ func (mem *CListMempool) TxsAvailable() <-chan struct{} {
...
+			mem.logger.Error("notified txs available but mempool is empty!")
+			return
 		}
-	}
+		if mem.txsAvailable != nil && !mem.notifiedTxsAvailable {
+			// channel cap is 1, so this will send once
+			mem.notifiedTxsAvailable = true
+			time.Sleep(mem.config.Consensus.TimeoutCommit)
```
Comment (melekes) [inline mempool/clist_mempool.go]: why `TimeoutCommit` though and not some configurable interval (say 1s)?

File: mempool/clist_mempool.go (hunk 1 above)
Comment (b00f) [reply mempool/clist_mempool.go]: we need to have timeout for mempool before sending notification. I think commit-timeout is fair enough. 
I didn't want to add new item to the mempool config. and I used the consensus config. It makes sense.

File: mempool/clist_mempool.go (hunk 2 above)
Comment (b00f) [reply mempool/clist_mempool.go]: Oh no panic please. Who knows what happens in threading apps!

File: mempool/clist_mempool.go (hunk 1 above)
Comment (b00f) [reply mempool/clist_mempool.go]: In my tests I never faced this condition.

File: mempool/clist_mempool.go (hunk 3 above)
Comment (b00f) [reply mempool/clist_mempool.go]: if 1 sec, then we create a block each second. But in consensus engine commit time out is set to 5 (default). I believe commit timeout should be same regardless of create_empty_block flag is set or not.

Comment (melekes) [reply mempool/clist_mempool.go]: I am not talking about changing TimeoutCommit. But rather using another variable

Comment (b00f) [reply mempool/clist_mempool.go]: You mean adding new item to mempool config?

Comment (melekes) [reply mempool/clist_mempool.go]: Yeah

Comment (b00f) [reply mempool/clist_mempool.go]: `no_empty_blocks_wait`, Is it good name for this config? Any suggestion?

Comment (b00f) [reply mempool/clist_mempool.go]: Another way, What's your idea I pass `TimeoutCommit` an new param to the constructor? So I don't need to pass whole config here.

**Comment by melekes** [comment]
Closing in favor of https://github.com/tendermint/tendermint/pull/3928

**Pull request closed without merging**